Added REPLACE as an alternative to INSERT.

sayore opened the pull request:
Replace effectivly looks for PRIMARY or UNIQUE constraints, if something matched it gets replaced, otherwise a new row will be added to the table. 
[Ref SQLite Replace](https://www.sqlite.org/lang_replace.html)

Commit:
Added REPLACE as an alternative to INSERT.

rahmanfadhil reviewed (changes requested):
Thanks for your contribution! You've put a lot of work on this one!

This feature looks promising, and the tests are solid. However, there are a few points you need to address before I'm ready to merge this PR.

rahmanfadhil commented on src/querybuilder.ts:
```diff
@@ -134,6 +135,23 @@ export class QueryBuilder {
     return this;
   }
 
+  public replace(data: QueryValues): QueryBuilder {
```
Please add a TSDoc comment to this method.

Example:

```
/**
 * Perform `REPLACE` query to the table.
 * 
 * It will look for `PRIMARY` and `UNIQUE` constraints.
 * If something matched, it gets removed from the table
 * and creates a new row with the given values.
 */
public replace(): QueryBuilder {
```

rahmanfadhil replied:
Please add a TSDoc comment to this method.

Example:

```
/**
 * Perform `REPLACE` query to the table.
 * 
 * It will look for `PRIMARY` and `UNIQUE` constraints.
 * If something matched, it gets removed from the table
 * and creates a new row with the given values.
 */
public replace(): QueryBuilder {
```

rahmanfadhil commented on src/querybuilder.ts:
```diff
@@ -381,6 +401,28 @@ export class QueryBuilder {
   }
 
+  /**
+   * Generate `INSERT` query string
```
Please update the TSDoc comment on this method.

```
/**
 * Generate `REPLACE` query string
 */
```

rahmanfadhil replied:
Please update the TSDoc comment on this method.

```
/**
 * Generate `REPLACE` query string
 */
```

rahmanfadhil commented on src/querybuilder.ts:
```diff
@@ -381,6 +401,28 @@ export class QueryBuilder {
+      const values = `(${Object.values(this.description.values).join(", ")})`;
+      query.push(fields, "VALUES", values);
+    } else {
+      throw new Error("Cannot perform insert query without values!");
```
Please update the error message in `toReplaceSQL` method.

```
throw new Error("Cannot perform replace query without values!");
```

Also, can you add a test that will catch this error if the user calls this method without any values?

Example:

```ts
Deno.test("QueryBuilder: throw an error if the `replace` method gets called without any values", () => {
  assertThrows(
    () => {
      (new QueryBuilder("users") as any).replace().toSQL();
    },
    Error,
    "Cannot perform replace query without values!",
  );
})
```

rahmanfadhil commented on src/querybuilder.ts:
```diff
@@ -381,6 +401,28 @@ export class QueryBuilder {
+      const values = `(${Object.values(this.description.values).join(", ")})`;
+      query.push(fields, "VALUES", values);
+    } else {
+      throw new Error("Cannot perform insert query without values!");
```
Please update the error message in toReplaceSQL method.

```
throw new Error("Cannot perform replace query without values!");
```

Also, can you add a test that will catch this error if the user calls this method without any values?

Example:

```ts
Deno.test("QueryBuilder: throw an error if the `replace` method gets called without any values", () => {
  assertThrows(
    () => {
      (new QueryBuilder("users") as any).replace().toSQL();
    },
    Error,
    "Cannot perform replace query without values!",
  );
})
```

sayore replied:
We probably can't catch the TypeError of the undefined at toReplaceSQL(), but we can check if the Object has no values at all. I did add a check to that now. And i did add a TypeError check too.

Commit:
- Fixed some comments
- Added JSDoc to replace() in QueryBuilder
-Added Tests for undefined | no Values for replace()

sayore commented:
Everything should be dandy now i think!
First PR ever so if i did miss any steps that's probably the reason. 👍

rahmanfadhil reviewed (approved):
Looks great!

Commit:
Fixed Formatting

rahmanfadhil commented:
I think I made some mistakes here when trying to resolve the conflicts. Can you create a branch called `querybuilder-replace` on your repository based on your `master` branch?

sayore commented:
Yeah i did and i think it's just 2 spaces in the comments. I pushed a commit into the branch too to fix this.

rahmanfadhil commented:
Can you resolve the conflicts in the `querybuilder_test.ts` file? I can't resolve this conflict with GitHub's web editor.

rahmanfadhil commented:
I don't understand...

You created this commit but I can't see the changes in this PR 🤔

<img width="304" alt="Screen Shot 2020-06-08 at 11 00 43" src="https://user-images.githubusercontent.com/28192207/83991592-6b9ee800-a977-11ea-8e12-28c7856839c8.png">

sayore commented:
well either github is a slow poke, or i can just open a new pr in a few minutes; maybe i need to re-request for a review 🤔

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

sayore commented:
well its merged i was the slowpoke 😅

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

rahmanfadhil commented:
Nevermind, it's merged now.

Thanks for your contribution!
